plasmapy/utils/decorators/checks.py

```diff
@@ -1,20 +1,1083 @@
+                              inspect.Parameter.POSITIONAL_ONLY,
+                              annotation=bound_args.signature.return_annotation)
+        for param in things_to_check.values():
+            # variable arguments are NOT checked
+            # e.g. in foo(x, y, *args, d=None, **kwargs) variable arguments
+            #      *args and **kwargs will NOT be checked
+            #
+            if param.kind in (inspect.Parameter.VAR_KEYWORD,
+                              inspect.Parameter.VAR_POSITIONAL):
+                continue
+
+            # grab the checks dictionary for the desired parameter
+            try:
+                param_in_checks = self.checks[param.name]
+            except KeyError:
+                # checks for parameter not specified
+                continue
+
+            # build `out_checks`
+            # read checks and/or apply defaults values
+            out_checks[param.name] = {}
+            for v_name, v_default in self.__check_defaults.items():
+                try:
+                    out_checks[param.name][v_name] = \
+                        param_in_checks.get(v_name, v_default)
+                except AttributeError:
+                    # for the case that checks are defined for an argument,
+                    # but is NOT a dictionary
+                    # (e.g. CheckValues(x=u.cm) ... this scenario could happen
+                    # curing subclassing)
+                    out_checks[param.name][v_name] = v_default
+
+        # Does `self.checks` indicate arguments not used by f?
+        missing_params = [
+            param
+            for param in set(self.checks.keys()) - set(out_checks.keys())
+        ]
+        if len(missing_params) > 0:
+            params_str = ", ".join(missing_params)
+            warnings.warn(PlasmaPyWarning(
+                f"Expected to value check parameters {params_str} but they "
+                f"are missing from the call to {self.f.__name__}"))
+
+        return out_checks
+
+    def _check_value(self, arg, arg_name: str, arg_checks: Dict[str, bool]):
+        """
+        Perform checks `arg_checks` on function argument `arg`.
+
+        Parameters
+        ----------
+        arg
+            The argument to be checked
+
+        arg_name: str
+            The name of the argument to be checked
+
+        arg_checks: Dict[str, bool]
+            The requested checks for the argument
+
+        Raises
+        ------
+        ValueError
+            raised if a check fails
+
+        """
+        if arg_name == 'checks_on_return':
+            valueerror_msg = f"The return value "
+        else:
+            valueerror_msg = f"The argument '{arg_name}' "
+        valueerror_msg += f"to function {self.f.__name__}() can not contain"
+
+        # check values
+        for ckey in self.__check_defaults:
+            if ckey == 'none_shall_pass':
+                if arg is None and arg_checks[ckey]:
+                    return
```

Comment:
This should probably be a `break` instead of a `return`.  If the argument is `None` and `none_shall_pass=True`, then there is no point in doing the remaining checks.  `return` and `break` essentially do the same thing here, but, when `return` is used, we'd run into issues of unaccessible code if there was code after the for-loop.